core/rapid_test.go

```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
 
 			// Wait until Quorum nodes finish their run loop
 			ctx, cancelFn := context.WithTimeout(context.Background(), time.Second*5)
-			if err := cluster.awaitNCompletions(ctx, int64(quorum(numNodes))); err != nil {
-				t.Fatalf(
-					fmt.Sprintf(
-						"unable to wait for nodes to complete, %v",
-						err,
-					),
+			err := cluster.awaitNCompletions(ctx, int64(quorum(numNodes)))
+			require.NoError(t, err, "unable to wait for nodes to complete")
+
+			// Shutdown the remaining nodes that might be hanging
+			cluster.forceShutdown()
+			cancelFn()
+		}
+
+		// Make sure that the inserted proposal is valid for each height
+		for _, proposalMap := range insertedProposals.proposals {
+			for _, insertedProposal := range proposalMap {
+				assert.Equal(t, message.proposal, insertedProposal)
+			}
+		}
+	})
+}
+
+// TestProperty_MajorityHonestNodes_BadProposal is a property-based test
```

Comment:
The malicious validator is always malicious, don't it? Could we do rather arbitrary? Particular rounds and stages?

Reply:
Yes, I will, but I would make it work first. Currently, the consensus can't be reached even with only one malicious node which is weird. Trying to figure out why it happens but still stuck here.